### ISSUE TITLE
Unit tests should be fast 🐎

### ISSUE THREAD

**Issue by jrieken** [microsoft/vscode#149712]
This is a continuation of https://github.com/microsoft/vscode/issues/134972. A unit test shouldn't take longer than 50ms. An easy way to "speed up" setTimeout and friends is @hediet's [`runWithFakedTimers`](https://github.com/microsoft/vscode/blob/b88626977819c490fa22e5b88eaec0f251aa97d6/src/vs/base/test/common/timeTravelScheduler.ts#L226)-utility

To reproduce these numbers

* run `./scripts/test.sh --reporter json > ~/Desktop/test_out.json`
* open the file and remove the first non-json lines
* run `jq '.tests | map(select(.duration > 50)) | sort_by(.duration) | [.[] | "- [ ] " + (.duration|tostring) + "ms, " + .fullTitle ]' ~/Desktop/test_out.json`

#### tests slower than 50ms

- [x] 51ms, MarkdownRenderer Code block renderer asyncRenderCallback should not be invoked if result is immediately disposed @mjbvz 
- [x] 104ms, MarkdownRenderer Code block renderer asyncRenderCallback should not be invoked if dispose is called before code block is rendered @mjbvz 
- [x] 55ms, ExtHostAuthentication Can get multiple sessions (with different scopes) in one extension @TylerLeonhardt 
- [x] 55ms, ExtHostAuthentication Can get multiple sessions (from different providers) in one extension @TylerLeonhardt 
- [x] 60ms, ExtHostAuthentication clearSessionPreference - true @TylerLeonhardt 
- [x] 72ms, ExtHostAuthentication silently getting session should return a session (if any) regardless of preference - fixes #137819 @TylerLeonhardt 
- [x] 93ms, Editor Controller issue #23913: Greater than 1000+ multi cursor typing replacement text appears inverted, lines begin to drop off selection @alexdima  
- [x] 51ms, Notebook Find Update find matches basics 3 @rebornix 
- [x] 53ms, Notebook Find Update find matches basics @rebornix 
- [x] 51ms, Notebook Find Update find matches, #112748 @rebornix

**Comment by sandy081** [comment]
Thanks @alexdima - for fixing tests in UserDataAutoSyncService

**Comment by sandy081** [comment]
Made IndexedDB tests faster but not fast enough to finish within less than 50ms. Because the bottle neck is IndexedDB. Any write transaction in IndexedDB is taking ~20ms and couple of tests need at least 2-3 write transactions. Hence checking off IndexedDB tests

**Comment by Tyriar** [comment]
@sandy081 I think tests like that which still take a long time should be renamed to `.integrationTest.ts` https://github.com/microsoft/vscode/issues/134972#issuecomment-942292889

**Comment by sandy081** [comment]
@jrieken Any reason I was assigned back?

**Comment by jrieken** [comment]
Sorry, my bad @sandy081

**Comment by sandy081** [comment]
@jrieken By any chance you overwrote the check marks?

**Comment by jrieken** [comment]
@sandy081 These are new checkboxes, everything slower than 125ms - I copied the list from the collapsed section

**Comment by sandy081** [comment]
@jrieken 

Sorry that I am confused. Following tests are fixed before and they got unchecked and I do not think they are slower. 

`UserDataAutoSyncService` tests were fixed before [here](https://github.com/microsoft/vscode/commit/68f3a6dd52e7a8fb55fcc381a4a447ba9978fe8c) and `IndexedDBFileSystemProvider` tests are moved to integration tests [here](https://github.com/microsoft/vscode/commit/4f7f2c6135bab4bbe67717aaf8f1c9ca5fb2bc4c). You can see the updated results when I run locally: `UserDataAutoSyncService` tests are not taking more than 50ms and `IndexedDBFileSystemProvider` tests do not exist.

<img width="1134" alt="image" src="https://user-images.githubusercontent.com/10746682/170837610-6559bb5b-f1d2-408d-bad4-462ccb06ce27.png">

I am not sure if I am missing anything here 😕

**Comment by sandy081** [comment]
Its also weird that I do not see any collapsed sections in my browser

<img width="976" alt="image" src="https://user-images.githubusercontent.com/10746682/170837906-754ba3d1-fefd-4e68-bd6b-85778892b6bf.png">

**Comment by alexr00** [comment]
I read through and couldn't find any of my tests. I think this is a case of "wrong Alex".

**Comment by bpasero** [comment]
Some observations after having looked into mine:
* in some of my slow tests I tend to do many assertions in a single test as opposed to doing 1 assertion per kind, as such I now split these tests up into smaller chunks
* in some tests I was able to squeeze out a bit more `ms` by doing `Promise.all` instead of awaiting promises one by one

I attribute slowness in my tests mainly to the fact that:
* I access the disk (because I actually want to test node.js APIs or native behaviour)
* I access a native library such as SQLite

I was trying out the fake timers thing, but obviously that does not help a lot for disk access or native libraries.

**Bottom line:** if we want unit tests to finish in `1-10ms` range then I think we need to pull out all unit tests that test disk or native libraries into a new suite that is either integration tests or can be controlled by some flags. I am not a big fan of integration tests mainly for the reason that I cannot run them as nice as I can with our test runner support (I brought this up in our standup to @connor4312).

Thoughts?

**Comment by jrieken** [comment]
@bpasero One thing we noticed is that quite some tests use the `DiskFileSystemProvider`. Those will surely benefit from an InMem-provider

<img width="944" alt="Screenshot 2022-06-09 at 15 21 59" src="https://user-images.githubusercontent.com/1794099/172857357-1824af67-756c-42d8-a462-4546c03d5d90.png">

**Comment by jrieken** [comment]
fyi - I reran the data extraction and updated the main comment here

**Comment by bpasero** [comment]
Pushed a couple of things with https://github.com/microsoft/vscode/pull/155147
* more adoption of `runWithFakedTimers` 🚀 
* figured out that we run a `fs.fdatasyncSync(fd);` for every `writeFile` to ensure data is flushed to disk which makes writing slower and figured out a way to disable this for most tests except the ones that want to test writing specifically
* adopted `inMemory` scheme for some of my tests and even found a bug in our in-memory fs-provider

**Before:**
`scripts/test.sh  23.83s user 5.24s system 41% cpu 1:10.09 total`

**After:**
`scripts/test.sh  24.31s user 5.06s system 47% cpu 1:02.10 total`

**Comment by bpasero** [comment]
Maybe after my PR landed, the list could be regenerated.

**Comment by bpasero** [comment]
Mine will be addressed by moving to integration via https://github.com/microsoft/vscode/pull/176011

I found new ones in notebook land, adding @rebornix 

I have asked @connor4312 to support our selfhost test extension showing test decorators in `xy.integrationTest.ts` files because one 🩸 papercut I have is that I can no longer use the selfhost test extension in those test files.

**Comment by Tyriar** [comment]
@aiday-mar we should action these as part of this issue:

<img width="837" alt="image" src="https://user-images.githubusercontent.com/2193314/227655169-83c8f92f-308a-491a-a63e-7d4c8b68d30f.png">

**Comment by azuregogo** [comment]
Excellent

في الجمعة، ١٢ يناير ٢٠٢٤، ٣:٢٧ ص Matt Bierner ***@***.***>
كتب:
